### PR TITLE
Add introduction

### LINKED ISSUE #4: Introduction is empty
Could you just copy the first three paragraphs from your recent blog post?

http://www.ncameron.org/blog/design-patterns-in-rust/

### PR DESCRIPTION
**Pull request by simonsan**
Reviving the ghost PR #43 

Regarding `functional` and `refactoring` folders:

Something that we might add to the `issues`:
> I was thinking something similar to what exists in the `patterns` folder, probably with an altered template. So far my ideas are:
> 
>     * `functional`
>       
>       * currying
>       * higher-order functions
>       * collections
> 
>     * `refactoring`
>       
>       * refactoring from `unwrap` to `Result`
>       * refactoring from `enum` to `Option`

Some point may still stand? Need to check.

2021-01-01 15_00_03-Window-Screenshot

Fixes #4

### CODE REVIEW

File: refactoring/README.md
Comment (MarcoIeni) [inline refactoring/README.md:8]:
```suggestion
> Shortcuts make for long days.

We can also use [idioms](idioms/README.md) to structure our code in a way that is understandable. 
```

File: patterns/README.md
Comment (MarcoIeni) [inline patterns/README.md:20]:
```suggestion
> The best code I ever wrote was code I never wrote.

If we apply YAGNI to design patterns, we see that the features of Rust allow us to throw out many patterns. For instance, there is no need for the [strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern) in Rust because we can just use [traits](https://doc.rust-lang.org/book/traits.html).
```

File: intro.md
Comment (MarcoIeni) [inline intro.md:8]: I think the reason behind the line 

```
Why Rust is a bit special - functional elements, type system - borrow checker
```

is that we should say that rust is not object oriented, so design patterns are different with respect to other object oriented languages.
No need to address this in this PR, but why might want to write this as a TODO.

File: patterns/README.md
Comment (MarcoIeni) [inline patterns/README.md:9]:
```suggestion
## Design patterns in Rust
```

File: patterns/README.md
Comment (MarcoIeni) [inline patterns/README.md:13]:
```suggestion
## YAGNI
```

File: rust_rocks.md
Comment (MarcoIeni) [inline rust_rocks.md:5]:
```suggestion
## Types
```

File: rust_rocks.md
Comment (MarcoIeni) [inline rust_rocks.md:11]:
```suggestion
## Ownership
```

File: intro.md
Comment (simonsan) [inline intro.md:8]: Not sure why the review comments go away, but I've added a sentence here in reference to what you've recommended

File: refactoring/README.md
Comment (MarcoIeni) [inline refactoring/README.md:1]: I've read this file and it doesn't seem to add any valuable information.
We could move individual sentences related to individual arguments in the introductions of the relevant chapters if you think they are relevant and missing.

For example in the `idioms` introduction we can write something like:

```
We can use idioms to structure our code in a way that is understandable
```

if you think it would fit well.

Comment (simonsan) [reply refactoring/README.md:1]: I think the person wanted to do more in these regards and I think it's quite useful, it's the content that will be added in the future that will make it more useful, for sure. Same for functional.

```
* `functional`
  
  * currying
  * higher-order functions
  * collections

* `refactoring`
  
  * refactoring from `unwrap` to `Result`
  * refactoring from `enum` to `Option`
  ```

Comment (simonsan) [reply refactoring/README.md:1]: I've added this list to the Readme file as a TODO with refactoring and functional as new points

Comment (simonsan) [reply refactoring/README.md:1]: So yeah, I agree with you that at this point it doesn't add too much valuable information, but I think it's nice to have as kind of a "placeholder" for the structure and to show people that there could be something added/PRed

File: intro.md
Comment (MarcoIeni) [reply intro.md:8]: yes!

File: README.md
Comment (MarcoIeni) [inline README.md:83]: Instead of putting this in the readme, why we don't place them in proper issues so we can have organized discussions about them? In this way we have a cleaner readme.

File: README.md
Comment (MarcoIeni) [inline README.md:9]:
```suggestion
An open source book about design patterns and idioms in the Rust programming
language that you can read [here](https://rust-unofficial.github.io/patterns/).

```

File: anti_patterns/README.md
Comment (MarcoIeni) [inline anti_patterns/README.md:6]:
```suggestion
Anti-patterns are not confined to code. For example, a process can be an anti-pattern, too.
```

File: refactoring/README.md
Comment (MarcoIeni) [reply refactoring/README.md:1]: I prefer to have a clean version of the book on master.
If it's just a phrase like "TODO talk also about this" it's fine, but if it's an entire chapter that has to be written it's too much I think. For the reader it feels like the book is incomplete.
I prefer to open an issue to track those missing chapters, since those are no more that ideas or "future requests" at the moment.

So basically I would just totally remove this file from this PR, again, moving some phrases around if you think they might be valuable somewhere else.
What do you think?

File: functional/README.md
Comment (MarcoIeni) [inline functional/README.md:53]:
```suggestion
| `a` | `b` | result |
| --- | --- | ------ |
| 0   | 1   | 1      |
| 1   | 2   | 3      |
| 3   | 3   | 6      |
| 6   | 4   | 10     |
| 10  | 5   | 15     |
| 15  | 6   | 21     |
| 21  | 7   | 28     |
| 28  | 8   | 36     |
| 36  | 9   | 45     |
| 45  | 10  | 55     |
```

File: functional/README.md
Comment (MarcoIeni) [inline functional/README.md:17]:
```suggestion
| `i` | `sum` |
```

File: functional/README.md
Comment (MarcoIeni) [inline functional/README.md:28]:
```suggestion
| `i` | `sum` |
| --- | ----- |
| 1   | 1     |
| 2   | 3     |
| 3   | 6     |
| 4   | 10    |
| 5   | 15    |
| 6   | 21    |
| 7   | 28    |
| 8   | 36    |
| 9   | 45    |
| 10  | 55    |
```

File: idioms/README.md
Comment (MarcoIeni) [inline idioms/README.md:5]:
```suggestion
The computer understands the machine code that is generated by the compiler. The language is therefore mostly beneficial to the developer. So, since we have this abstraction layer, why not put it to good use and make it simple?
Remember the KISS principle: "Keep It Simple, Stupid".
```

Comment (MarcoIeni) [reply idioms/README.md:5]: Maybe I would totally delete the phrase "Remember the KISS principle: "Keep It Simple, Stupid".
I like the kiss principle, but that single phrase seems a little bit out of context here. What do you think?

File: rust_rocks.md
Comment (MarcoIeni) [inline rust_rocks.md:3]: The question is: shall we do it?

File: rust_rocks.md
Comment (MarcoIeni) [inline rust_rocks.md:13]: Yes, it's true, but this chapter without a proper explanation and examples doesn't add any value..

File: rust_rocks.md
Comment (MarcoIeni) [inline rust_rocks.md:9]: This seems like a good intro to the new type pattern: https://rust-unofficial.github.io/patterns/patterns/newtype.html

File: README.md
Comment (simonsan) [reply README.md:83]: I would propose to do both actually. An link in the README.md to the issues. In this way we get attention to the issue itself on the "frontpage" of the repository, which I like and might convince people to commit more patterns. I personally don't have a big problem with the README.md being not that clean, now that we have the book rendering going on - which is the cleanest we probably can get I feel. What do you think about that?

Comment (simonsan) [reply README.md:83]: The linking of the issues in the README.md on TODOs I would prefer to do when we have the PRs merged, so we have a better overview and can use a single PR to link all issues.

File: refactoring/README.md
Comment (simonsan) [reply refactoring/README.md:1]: Hmm, I understand where you coming from and I partly agree for sure. On the other hand not every chapter in the repository needs to be rendered to HTML for the book. We can influence that with the SUMMARY.md file. So we could just not link these files in the book but have them in the repository already. Personally I would prefer this solution.

File: idioms/README.md
Comment (simonsan) [reply idioms/README.md:5]: Maybe that gives people that are not familiar with the KISS principle some small input to Google it and read it up themselves? Not sure though. Maybe another small phrase describing why it's important like an explanatory cliffhanger could be good then?

File: rust_rocks.md
Comment (simonsan) [reply rust_rocks.md:3]: I was unsure as well. And also didn't feel it's needed, but this file I feel the author wanted to use to introduce Rust to the general public as in `explain what it can do, it's limitations, borrowing, etc.`. But yeah, really not sure. :-/

Comment (simonsan) [reply rust_rocks.md:3]: But maybe I would remove this file from the PR, factor out things that could be nice in other files as your stated with `Types` and then we discuss about a small introduction to Rust itself in another PR/discussion/issue?

File: rust_rocks.md
Comment (simonsan) [reply rust_rocks.md:13]: Same here: "But maybe I would remove this file from the PR, factor out things that could be nice in other files as you stated with `Types` and then we discuss about a small introduction to Rust itself in another PR/discussion/issue?`

File: refactoring/README.md
Comment (simonsan) [reply refactoring/README.md:1]: I removed the link for refactoring-README.md from SUMMARY.md so it is marked as a draft in that regards (it will be greyed out in the book). Is that fine with you or would you rather completely remove the Headline from the TOC?

File: idioms/README.md
Comment (simonsan) [reply idioms/README.md]: Added a small claim of the KISS principle to make clearer. What do you think, still to be removed?

File: rust_rocks.md
Comment (simonsan) [reply rust_rocks.md:9]: Done

File: rust_rocks.md
Comment (simonsan) [reply rust_rocks.md:13]: Removed rust_rocks.md

File: rust_rocks.md
Comment (simonsan) [reply rust_rocks.md:3]: Removed rust_rocks.md and used introduction sentence in newtype.md

File: refactoring/README.md
Comment (simonsan) [reply refactoring/README.md:1]: 2021-01-02 05_58_46-Functional Programming - Rust Design Patterns – Mozilla Firefox-Screenshot

Comment (simonsan) [reply refactoring/README.md:1]: Like this it is now, which imho is fine. It's not a link.

2021-01-02 06_01_18-Introduction - Rust Design Patterns – Mozilla Firefox-Screenshot

File: patterns/newtype.md
Comment (MarcoIeni) [inline patterns/newtype.md:4]:
```suggestion

A really convenient application of the Rust type system is the Newtype pattern.

```

File: README.md
Comment (MarcoIeni) [reply README.md:79]: Sure, missing patterns could be linked in the README!
Anyway once you merge this in the README, you have to explain the details in the issue. I don't know what the contributor meant with "refactoring from enum to Option" for example.
Maybe we should first create the detailed issue and once we agree that it should be done and explain what we want, we can link it to the README. What do you think?

Comment (simonsan) [reply README.md:79]: sounds good

File: refactoring/README.md
Comment (MarcoIeni) [reply refactoring/README.md:1]: Honestly it looks really ugly :rofl:
If I were a reader, after I see this I would submit a PR to fix the link :sweat_smile: 
I mean..we don't even have a plan on what to put in this "Refactoring" section I think.
This file only contains a summary of the rest of the book :/
Let's open an issue to discuss *if* we want a refactoring chapter and what should be discussed there before adding a draft chapter in the book which will remain in this state for a long time I presume.

File: idioms/README.md
Comment (MarcoIeni) [reply idioms/README.md]: It's really nice now, thanks!

File: refactoring/README.md
Comment (simonsan) [reply refactoring/README.md:1]: Ok, i'll take it out of the summary then 🤣

Comment (MarcoIeni) [reply refactoring/README.md:1]: Should we commit this file in your opinion?

Comment (simonsan) [reply refactoring/README.md:1]: I know there is not much in it already, but I really think it is a good idea to consider refactoring as a related part of software design (->patterns). 

"Refactoring is intended to improve the design, structure, and/or implementation of the software (its non-functional attributes), while preserving its functionality."

I think it would be good to start putting it as we should see the Book itself as a clean surface, the repository structure should be clean as well, but I assume we can fix that soonish and it will not stay long so empty.

Comment (simonsan) [reply refactoring/README.md:1]: Also because there are `idoms` as well. I think the scope of this repository could really be around the process of good software design in Rust in the end.

File: intro.md
Comment (simonsan) [reply intro.md]: added it as a TODO in intro

File: refactoring/README.md
Comment (MarcoIeni) [reply refactoring/README.md:1]: I feel like idioms are more related to design patterns with respect to refactoring.
But if you feel like this topic should be added it's ok :)

File: README.md
Comment (MarcoIeni) [reply README.md:79]: So should we remove these items from the readme and create two isssues, one for functional programming and one for refactoring?

Comment (simonsan) [reply README.md]: done

Comment (simonsan) [reply README.md]: The relevant comment is here:
https://github.com/rust-unofficial/patterns/pull/43#issuecomment-258056968

**Review by MarcoIeni — APPROVED** [review]

**Comment by simonsan** [comment]
Nice!

File: functional/README.md
Comment (pickfire) [inline functional/README.md:28]: Shouldn't numbers be right-aligned to make it easier to read?

File: idioms/README.md
Comment (pickfire) [inline idioms/README.md:7]: This sounds like go to me. I feel like rust is aiming to be "keep it simple but not simpler". Go is able to be simple by having tons of footguns as a result. Rust I would say aims more to be correct than simple (this is why rust became so complicated), one good thing is that the footguns are reduced to bare minimum.

File: intro.md
Comment (pickfire) [inline intro.md:12]: Feels like the same wordings, maybe we should let readers know what are the differences between them? "Anti-patterns" is a tough english word, we should have more descriptions.

File: patterns/README.md
Comment (pickfire) [inline patterns/README.md:21]: I think we could remove this, I wonder if illustration is really needed here for understanding.

File: patterns/newtype.md
Comment (pickfire) [inline patterns/newtype.md:3]: How is this paragraph related to "Newtype" pattern?

**Comment by simonsan** [comment]
@pickfire It's hard to deal with that feedback in a closed and merged PR. What do you think about changing the things directly in these files to your liking and create a new PR where we can discuss about your proposed changes? Would make it much easier to follow for everyone I think. Happy new year! ;)

File: patterns/newtype.md
Comment (pickfire) [inline patterns/newtype.md:5]: How is Newtype pattern convenient? I think we should use a different word other than convenient or maybe we should explain why is it convenient. Or maybe we could just

> An application of the Rust type system is the Newtype pattern.

File: refactoring/README.md
Comment (pickfire) [inline refactoring/README.md:5]: We should link "DRY" for "Don't repeat yourself" wiki page.

File: functional/README.md
Comment (MarcoIeni) [reply functional/README.md:28]: yes, the online formatter I used was not that great. you can open a PR if you want.

File: idioms/README.md
Comment (MarcoIeni) [reply idioms/README.md:7]: I strongly agree with you.

We can say something like:
"even if the kiss principle is important rust tends to prefer correctness over simplicity."

If you want to do a PR where you explain this it would be nice I think (you don't have to report my words necessarily).

File: patterns/newtype.md
Comment (MarcoIeni) [reply patterns/newtype.md:3]: I thought it might be a nice "intro" to the pattern 😅
If you don't like it we can discuss about it.
Anyway as simonsan said maybe it is better to open a new PR.

File: patterns/newtype.md
Comment (MarcoIeni) [reply patterns/newtype.md:5]: > What if in some cases we want to have a type to behave similarly but still different to another type?

I like this :)

File: refactoring/README.md
Comment (MarcoIeni) [reply refactoring/README.md:5]: definitely!

**Comment by pickfire** [comment]
> @pickfire It's hard to deal with that feedback in a closed and merged PR. What do you think about changing the things directly in these files to your liking and create a new PR where we can discuss about your proposed changes? Would make it much easier to follow for everyone I think. Happy new year! ;)

Happy new year!

I wish there is more time gap since I am interested to help review these too (of course only those that I am able to). I see that by the time I read it (within 1 or 2 day time) it was already merged. @simonsan How did you get to be the maintainer here? I remember that it wasn't active for a while and then you suddenly come and change.

**Comment by simonsan** [comment]
> I wish there is more time gap since I am interested to help review these too (of course only those that I am able to). I see that by the time I read it (within 1 or 2 day time) it was already merged.

This PR was a ghost PR from October 2016, most of the PRs, beside the repository structure we are changing like setting up CI or creating a book, were relatively old PRs. We want to work down the list and create a great learning resource for the community with it because imho this is excellent content and it will get even better with more people participating in the future. The other left over PRs are much bigger topics than the ones that were merged, so there will be a timegap for sure. I'm thankful as well, if there are more eyes on it with reviews, please feel free to review them as well. The velocity will slow down soonish, as we handled most of the open PRs already.

> @simonsan How did you get to be the maintainer here? I remember that it wasn't active for a while and then you suddenly come and change.

I think this is a good collection of what happened and an actual status:
https://www.reddit.com/r/rust/comments/kowtqn/rust_design_patterns_now_also_as_a_book/

Cheers ;-)

**Comment by MarcoIeni** [comment]
Pickfire, feel free to leave a comment on a pr if you want to review it and we will wait for your review, even for different days, no problem! 
Right now I'm overwhelmed by github notifications, so any help is HIGHLY appreciated! :)

**Pull request merged**
